src/main/java/io/leonis/zosma/ipc/peripheral/ControllerPublisher.java

```diff
@@ -1,87 +1,86 @@
+   * Creates an ControllerPublisher that allows for dynamic rebinding based on the latest map
+   * published on the provided {@code mapPublisher}.
    *
-   * @param interval Duration between polls.
-   * @param adapter Adapter used to transform jamepad state.
-   * @param amount Amount of controllers to listen for.
+   * @param amount       The amount of controllers to listen for.
+   * @param interval     The poll rate.
+   * @param adapter      The adapter from jamepad to an controller object.
+   * @param mapPublisher Publisher of the assignment map.
    */
   public ControllerPublisher(
+      final int amount,
       final Duration interval,
       final Function<ControllerIndex, C> adapter,
-      final int amount
-  ) {
-    this.manager = new ControllerManager(amount);
-    this.manager.initSDLGamepad();
-    this.interval = interval;
-    this.adapter = adapter;
+      final Publisher<Controller.MapSupplier<I>> mapPublisher
+  ){
+    this(new ControllerSetPublisher<>(interval, adapter, amount), mapPublisher);
   }
 
   /**
-   * {@inheritDoc}
+   * Creates an ControllerPublisher with a static mapping.
+   *
+   * @param amount   The amount of controllers to listen for.
+   * @param interval The poll rate.
+   * @param adapter  The adapter from jamepad to an controller object.
+   * @param mapping  The assignment mapping.
    */
+  public ControllerPublisher(
+      final int amount,
+      final Duration interval,
+      final Function<ControllerIndex, C> adapter,
+      final Map<I, Set<ControllerIdentity>> mapping
+  ){
+    this(new ControllerSetPublisher<>(interval, adapter, amount),
+         Flux.just(() -> mapping));
+  }
+
   @Override
-  public void subscribe(final Subscriber<? super Controller.SetSupplier<C>> subscriber) {
-    Flux.interval(interval)
-        .map(tick ->
-            // Loop all controllers
-            IntStream.range(0, manager.getNumControllers())
-                // grab the states
-                .mapToObj(manager::getControllerIndex)
-                // ignore any disconnected controller
-                .filter(ControllerIndex::isConnected)
-                // map to custom state and collect as a set.
-                .map(adapter)
-                .collect(Collectors.toSet()))
-        .map(Frame::new)
-        .subscribe(subscriber);
+  public void subscribe(final Subscriber<? super Frame<C, I>> s) {
+    Flux.combineLatest(setSupplierPublisher, mapSupplierPublisher, Frame::new).subscribe(s);
   }
 
-  /**
-   * The Class Frame.
-   *
-   * @author Jeroen de Jong
-   */
-  @Value
-  private class Frame implements Controller.SetSupplier<C> {
-    /**
-     * The {@link Set} of {@link Controller controllers} captured in this frame.
-     */
-    private Set<C> controllerSet;
+  @AllArgsConstructor
+  static final class Frame<C extends Controller, I extends Identity>
+      implements Controller.SetSupplier<C>, Controller.MapSupplier<I> {
+    @lombok.experimental.Delegate
```

Comment:
these should be `@Delegate`

Reply:
Can't be done due to the import of `lombok.*` or something. Please check for yourself locally whether you can get it to work though :P